src/protocol/sort/shuffle.rs

```diff
@@ -1,99 +1,226 @@
 use crate::helpers::prss::PrssSpace;
-use crate::helpers::Direction;
+use permutation::Permutation;
 use rand::seq::SliceRandom;
 use rand::SeedableRng;
 use rand_chacha::ChaCha8Rng;
 
+use embed_doc_image::embed_doc_image;
+use futures::future::try_join_all;
+
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::{
+        mesh::{Gateway, Mesh},
+        prss::SpaceIndex,
+        Direction, Identity,
+    },
+    protocol::{context::ProtocolContext, RecordId, Step},
+    secret_sharing::Replicated,
+};
+
+use super::{
+    apply::apply_inv,
+    reshare::Reshare,
+    ShuffleStep::{self, Step1, Step2, Step3},
+};
+
 #[allow(dead_code)]
-pub struct Shuffle {}
+pub struct Shuffle<'a, F, S> {
+    input: &'a mut Vec<Replicated<F>>,
+    step_fn: fn(ShuffleStep) -> S,
+}
 
-impl Shuffle {
-    /// This implements Fisher Yates shuffle described here <https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle>
+/// This is SHUFFLE(Algorithm 1) described in <https://eprint.iacr.org/2019/695.pdf>.
+/// This protocol shuffles the given inputs across 3 helpers making them indistinguishable to the helpers
+impl<'a, F: Field, S: Step + SpaceIndex> Shuffle<'a, F, S> {
     #[allow(dead_code)]
+    pub fn new(input: &'a mut Vec<Replicated<F>>, step_fn: fn(ShuffleStep) -> S) -> Self {
+        Self { input, step_fn }
+    }
+
+    #[allow(clippy::cast_possible_truncation)]
     fn generate_random_permutation(
         batchsize: usize,
         direction: Direction,
-        index: u128,
         prss: &PrssSpace,
-    ) -> Vec<usize> {
-        let mut permutation: Vec<usize> = (0..batchsize).collect();
-        let seed = prss.generate_values(index);
+    ) -> Permutation {
+        /// This implements Fisher Yates shuffle described here <https://en.wikipedia.org/wiki/Fisher%E2%80%93Yates_shuffle>
+        fn generate_random_permutation_internal(
```

Comment:
I really feel like defining this inline function makes this code much less readable. It took me a while to figure out what is going on.

Given that this function is called just once, I'd recommend NOT defining it this way, and instead just inlining the code. I think that'll save you the hassle of dealing with the two "seed" values. Right now there is a lot of code to just sort of write and read from these two values.

Reply:
We do need 2 seed values (since we get a seed of u128 from prss while chacha8rng needs a u256). This will change once @martinthomson finishes changes for PrssRng to support it. However, agreed on inlining this code to make it readable.

Reply:
Let me know if this looks better